CEP for Deprecation Policy (#27)
conda-incubator/ceps: cep-deprecation.md

<table>
<tr><td> Title </td><td> Conda Deprecation Schedule </td>
<tr><td> Status </td><td> Draft </td></tr>
<tr><td> Author(s) </td><td> Ken Odegard &lt;kodegard@anaconda.com&gt; </td></tr>
<tr><td> Created </td><td> May 20, 2022 </td></tr>
<tr><td> Updated </td><td> July 20, 2022 </td></tr>
<tr><td> Discussion </td><td> NA </td></tr>
<tr><td> Implementation </td><td> NA </td></tr>
</table>

## Abstract

> Describes a deprecation schedule to properly warn about upcoming removals from the codebase. This policy expands on ideas and terminology defined in the Conda Release Schedule (CEP-?).

Reviewer: ```suggestion
This CEP describes a deprecation schedule to properly warn about upcoming removals from the codebase. This policy expands on ideas and terminology defined in the Conda Release Schedule (CEP-?).
```

Reviewer: ☝️ suggestion gets rid of the sentence fragment, but I'm also curious which codebase this CEP applies to? Just the `conda` codebase? Or all of our projects?

Author: For now, the deprecation policy will only apply to projects that have adopted the bi-monthly release schedule. I'll update accordingly.

## Specification

We propose a deprecation schedule that is slower than the Conda Release Schedule (CEP-?). This is in acknowledgment of our diverse user groups (e.g. everything from per user per machine installs to multi-user installs on shared clusters).

> We define a new type of release, deprecation release, to augment the regular release. A deprecation release occurs three (3) times every year in January, May, and September:

Reviewer: ```suggestion
We define a new type of release, a deprecation release, to augment regular releases. A deprecation release occurs three (3) times every year in January, May, and September:
```

| Version | Release Type |
|---|---|
| `YY.1.0` | regular, **deprecation** |
| `YY.2.0` | optional |
| `YY.3.0` | regular |
| `YY.4.0` | optional |
| `YY.5.0` | regular, **deprecation** |
| `YY.6.0` | optional |
| `YY.7.0` | regular |
| `YY.8.0` | optional |
| `YY.9.0` | regular, **deprecation** |
| `YY.10.0` | optional |
| `YY.11.0` | regular |
| `YY.12.0` | optional |

All deprecations will need to transition through three (3) states:

> 1. **pending deprecation**: when a feature is marked for future deprecation

Reviewer: ```suggestion
1. **Pending deprecation**: when a feature is marked for future deprecation
```

>     - any feature can be marked as *pending deprecation* in **any release**

Reviewer: ```suggestion
    - Any feature can be marked as *pending deprecation* in **any release**
```

> 2. **deprecated**: when a feature is marked for future removal

Reviewer: ```suggestion
2. **Deprecated**: when a feature is marked for future removal
```

>     - all *pending deprecations* that have been through at least one (1+) **regular release** (3-6 months after being marked as *pending deprecation*) are relabeled as *deprecated* in the next **deprecation release**

Reviewer: ```suggestion
    - All *pending deprecations* that have been through at least one (1+) **regular release** (3-6 months after being marked as *pending deprecation*) are relabeled as *deprecated* in the next **deprecation release**
```

> 3. **removed**: when a feature is removed from the codebase

Reviewer: ```suggestion
3. **Removed**: when a feature is removed from the codebase
```

>     - all *deprecations* are removed in the next **deprecation release**

Reviewer: ```suggestion
    - All *deprecations* are removed in the next **deprecation release**
```

    Removed     :milestone, 2024-01

    section Jun
    Pending     :           2023-06, 2023-09
    Deprecated  :active,    2023-09, 2024-01
    Removed     :milestone, 2024-01

    section Jul
    Pending     :           2023-07, 2024-01
    Deprecated  :active,    2024-01, 2024-05
    Removed     :milestone, 2024-05

    section Aug
    Pending     :           2023-08, 2024-01
    Deprecated  :active,    2024-01, 2024-05
    Removed     :milestone, 2024-05
```

```mermaid
gantt
    dateFormat YYYY-MM
    axisFormat YY.%_m
    title      September-December Deprecations

    section Sep
    Pending     :           2023-09, 2024-01
    Deprecated  :active,    2024-01, 2024-05
    Removed     :milestone, 2024-05

    section Oct
    Pending     :           2023-10, 2024-01
    Deprecated  :active,    2024-01, 2024-05
    Removed     :milestone, 2024-05

    section Nov
    Pending     :           2023-11, 2024-05
    Deprecated  :active,    2024-05, 2024-09
    Removed     :milestone, 2024-09

    section Dec
    Pending     :           2023-12, 2024-05
    Deprecated  :active,    2024-05, 2024-09
    Removed     :milestone, 2024-09
```

> Occasionally there may be code changes that warrant a longer deprecation schedule. If that occurs the deprecation warning will clearly specify that a deviation is occurring and what the expected schedule will be instead.

Reviewer: ```suggestion
Occasionally, there may be changes to the codebase that warrant a longer deprecation schedule. If that occurs, the deprecation warning will clearly specify that a deviation is occurring and what the expected schedule will be instead.
```

## Motivation

> Help prevent unexpected breakage of downstream tooling as the codebase evolves.

Reviewer: ```suggestion
This CEP will help prevent unexpected breakage of downstream tooling as the codebase evolves.
```

## Backwards Compatibility

This is backwards compatible and will also encourage better backwards compatibility.

## Alternatives

1. Ad hoc deprecations.
   - Unpredictable and unreliable.

> 2. Mark as **pending deprecation** in one release, mark as a **deprecation** the next release, and **remove** in third release.

Reviewer: ```suggestion
2. Mark as **pending deprecation** in one release, mark as a **deprecated** in the next release, and **remove** in third release.
```

<table>
<tr><td> Title </td><td> Conda Deprecation Schedule </td>
<tr><td> Status </td><td> Draft </td></tr>
<tr><td> Author(s) </td><td> Ken Odegard &lt;kodegard@anaconda.com&gt; </td></tr>
<tr><td> Created </td><td> May 20, 2022 </td></tr>
<tr><td> Updated </td><td> July 20, 2022 </td></tr>
<tr><td> Discussion </td><td> NA </td></tr>
<tr><td> Implementation </td><td> NA </td></tr>
</table>

## Abstract

This CEP describes a deprecation schedule to properly warn about upcoming removals from the codebase. This policy expands on ideas and terminology defined in the Conda Release Schedule (CEP-?).

> **Note**
> This CEP is only applicable for projects that have adopted the Conda Release Schedule (CEP-?).

## Specification

We propose a deprecation schedule that is slower than the Conda Release Schedule (CEP-?). This is in acknowledgment of our diverse user groups (e.g. everything from per user per machine installs to multi-user installs on shared clusters).

> We define a new type of release, a deprecation release, to augment regular releases. A deprecation release occurs three (3) times every year in January, May, and September:
> 
> | Version | Release Type |
> |---|---|
> | `YY.1.0` | regular, **deprecation** |
> | `YY.2.0` | optional |
> | `YY.3.0` | regular |
> | `YY.4.0` | optional |
> | `YY.5.0` | regular, **deprecation** |
> | `YY.6.0` | optional |
> | `YY.7.0` | regular |
> | `YY.8.0` | optional |
> | `YY.9.0` | regular, **deprecation** |
> | `YY.10.0` | optional |
> | `YY.11.0` | regular |
> | `YY.12.0` | optional |

Reviewer 2: I think 3 times per year makes the deprecation cycle too fast for most users that only update conda irregular, if at all.

I suggest to do a deprecation cycle twice every year until we have encouraged users to update conda more actively.

I would use March and September since both are far enough from various holidays and seasonal slowdowns away to make upgrades easier to handle.

The only downside IMO would be that the actual removal of deprecated features would move around on the calendar, given the three-step default.

```suggestion
We define a new type of release, a deprecation release, to augment regular releases. A deprecation release occurs twice every year in March and September:

| Version | Release Type |
|---|---|
| `YY.1.0` | regular |
| `YY.2.0` | optional |
| `YY.3.0` | regular, **deprecation** |
| `YY.4.0` | optional |
| `YY.5.0` | regular |
| `YY.6.0` | optional |
| `YY.7.0` | regular |
| `YY.8.0` | optional |
| `YY.9.0` | regular, **deprecation** |
| `YY.10.0` | optional |
| `YY.11.0` | regular |
| `YY.12.0` | optional |
```

What do you think?

Reviewer 2: BTW pip's deprecation policy is less strict and just leaves it to the maintainers, but that may not be sensible given that conda is a larger project and it's good to default to clear and predictable schedules.

Author: You're absolutely right, I was not a fan of how this created a new 3 step schedule. Now that we've defined the concept of a deprecation release (instead of counting the number of releases), it makes a lot more sense as you suggested.
